src/mono/mono/mini/simd-intrinsics.c

```diff
@@ -396,11 +396,11 @@ emit_simd_ins_for_binary_op (MonoCompile *cfg, MonoClass *klass, MonoMethodSigna
 			case SN_op_Multiply: {
 				const char *class_name = m_class_get_name (klass);
 				if (strcmp ("Quaternion", class_name) && strcmp ("Plane", class_name)) {
-					if (!type_is_simd_vector (fsig->params [1]))
+					if (!type_is_simd_vector (fsig->params [1])) // vector * scalar
```

Comment:
nit: The checks used here and elsewhere aren't really consistent

Some places are only checking the second parameter is not SIMD, others are checking it is explicitly a primitive. Likewise other checks are validating that both parameters are vector, etc.

It would be nice to be consistent here and do the right checks in all scenarios. That is, we should have at least one vector type and for the cases that accept scalars, the scalar should be a supported primitive type (not all primitive types are supported).

The intrinsics that support scalars in either position (multiply) or only in one position (divide for the second parameter) should likely be explicitly separate as well. There's some code paths, like on L2067 below, where the handling for op1 to be scalar is also being done for MinNative, Subtract, and other intrinsics where it should never be valid.

Reply:
That's a good point. I refactored both `emit_simd_ins_for_binary_op` and the callsite in `emit_sri_vector` to check for scalars and vectors for respective APIs. Let me know what you think about the changes?